feature[input]: add Halo text area

donald-okara opened the pull request:
## Description
> information of what was being worked on (__if necessary__)

This PR introduces a text area component (filled and outlined)

## Notes
> additional information of what was being worked on like testing or something that changed (__if necessary__)

It also adds ui tooling so as one can check how the components works without having to run the app each time

## Issues
> issue number of what was being worked on (__if necessary__)

Closes #48 

None

## Screenshot

<p float="left">
  <img src="https://github.com/user-attachments/assets/ea67c8ed-fdfe-4124-aa9e-44251154c650" width="400" />
  <img src="https://github.com/user-attachments/assets/7bfa3ab8-49c9-476a-9766-a075ad89e9b9" width="400" />
  <img src="https://github.com/user-attachments/assets/c1dd3092-5ba0-4e71-9ac3-5c3429089f1b" width="400" />
</p>


Still WIP

## Reviewers
> who's going to review the PR

@janewaitara @ericwafula @MamboBryan

[//]: # (remove your name above if you're the one who's created the PR)

donald-okara committed:
feat: add ui tooling and preview dependencies

- This commit introduces dependencies for `androidx.compose.ui.tooling` and `androidx.compose.ui.tooling-preview` to enable UI previews.
- Additionally, a new file `TextAreasShowcase.kt` is added, containing a composable function with a `@Preview` annotation to demonstrate text area components.

donald-okara committed:
feat/text-area: add text area components

This commit introduces `HaloBaseTextArea` and `HaloFilledTextArea` composables for creating text areas.

It also adds a `TextAreaHeightMode` enum to control whether the text area expands or has a fixed height.

The showcase has been updated to include examples of the new text area components.

donald-okara committed:
feat/text-area : adds outlined text area

This commit adds the outlined text area component.

donald-okara committed:
fix: added base text area preview to showcase

donald-okara committed:
feat: add TextArea to gallery screen

This commit introduces the `TextArea` component to the gallery screen, with options for both `Outlined` and `Filled` variants.

donald-okara committed:
refactor(text-area): remove leading/trailing icons

This commit removes the `leading` and `trailing` icon parameters from `HaloBaseTextArea`, `HaloFilledTextArea`, and `HaloOutlinedTextArea`.

The internal layout of `HaloBaseTextArea` has been updated to use a `Box` instead of a `Row` for the main content area, simplifying the layout and removing the need for explicit leading/trailing icon placement.

donald-okara committed:
docs: update TextAreaShowcase KDoc

The KDoc for `TextAreaShowcase` has been updated to be more descriptive of its purpose.

donald-okara committed:
feat(text-area): add text area section to desktop showcase

This commit introduces a dedicated "Text Area" section to the desktop showcase application.

The `TextAreaSection.kt` file provides examples of:
- Filled and Outlined text areas in various states (enabled, disabled, error, read-only).
- Different height modes for text areas (Fixed and Expandable).

The `GalleryScreenModel.kt` and `GalleryScreen.kt` have been updated to reflect this new section.

Additionally, `TextAreaHeightMode` has been refactored from an enum to a sealed class, allowing for more specific configuration of fixed and expandable height modes (e.g., specifying the number of lines for fixed height).

donald-okara committed:
refactor(textarea): minor code cleanup and formatting

This commit includes the following changes:

- Removed unused imports from `TextAreaSection.kt`.
- Shortened the placeholder text in the desktop sample for `TextAreaSection.kt` to improve readability.
- Added trailing commas in `TextAreaHeightMode.kt` and `TextAreas.kt` for consistency.
- Improved code formatting in `TextAreas.kt` for better readability.
- Slightly adjusted padding in `TextAreaSection.kt` for the "Height mode" title.

ericwafula commented on halo/src/commonMain/kotlin/org/bizilabs/halo/base/TextAreaHeightMode.kt:
You can change this from a sealed class to a sealed interface since the parent class has an empty constructor

<img width="542" height="350" alt="image" src="https://github.com/user-attachments/assets/655d7567-8849-400f-842f-4caad7df0945" />

ericwafula commented on halo/src/commonMain/kotlin/org/bizilabs/halo/components/textarea/TextAreas.kt:
I don't think we need the singleLine variable here since we already have the TextAreaHeightMode which defines the height of the textField.
<img width="231" height="28" alt="image" src="https://github.com/user-attachments/assets/9239065d-e9f7-419c-a29c-afe222a675c7" />

ericwafula replied:
Also, prefer sticking with subject when expressions over condition when expressions since it makes it easier to read and understand what the code is doing. You can update the following and any other occurences.

<img width="630" height="217" alt="image" src="https://github.com/user-attachments/assets/ae2dadcf-b583-429f-a038-50782f25cd3b" />

You can use this as an example:

<img width="659" height="397" alt="image" src="https://github.com/user-attachments/assets/24fc3c70-461b-4919-a078-3d5cc7a14680" />

ericwafula replied:
I am finding it hard to understand the role of the helper parameter:

<img width="313" height="23" alt="image" src="https://github.com/user-attachments/assets/f81ee17d-e03c-4103-87b0-c318397ef16f" />

Maybe you can use documentation to communicate what it does better. You could also do the same for the count parameter

donald-okara replied:
Hello Eric. The documentation, i.e the issue opened and the desktop app, are not clear on what the helper does. I assumed it is important to the larger use case of textfields in Halo, so I kept it. However, I would like to know its purpose and whether it is necessary for the text area. Maybe @MamboBryan could weigh in on this, as he is the one who created the issue?

donald-okara committed:
refactor(text-area): convert TextAreaHeightMode to a sealed interface

This commit refactors `TextAreaHeightMode` from a `sealed class` to a `sealed interface`.

This change modernizes the code and removes the need for the `Fixed` and `Expandable` data classes to explicitly inherit from `TextAreaHeightMode()`.

donald-okara committed:
refactor(textarea): remove `singleLine` parameter

This commit removes the `singleLine` parameter from `HaloBaseTextArea`, `HaloFilledTextArea`, and `HaloOutlinedTextArea`.

Text areas are inherently multi-line, so this parameter was redundant and has been removed to simplify the API. The `maxLines` and `minLines` parameters should be used to control line behavior.

donald-okara committed:
refactor(textarea): simplify border color logic

This commit refactors the `when` expression that determines the border color for the text area component. The logic has been simplified by removing an unnecessary nested `if/else` block for the focused and default states, improving code readability.
